src/renderer/components/ImageCropper/index.tsx

```diff
@@ -187,49 +260,95 @@ export default function ImageCropper({
       )
     )
 
+    // clamp cursor position
     const nX = Math.max(
-      Math.min((imgW - targetWidth.current / zoom.current) / 2, x),
-      (targetWidth.current / zoom.current - imgW) / 2
+      Math.min(imgW - targetWidth.current / zoom.current / 2, x),
+      targetWidth.current / zoom.current / 2
     )
     const nY = Math.max(
-      Math.min((imgH - targetHeight.current / zoom.current) / 2, y),
-      (targetHeight.current / zoom.current - imgH) / 2
+      Math.min(imgH - targetHeight.current / zoom.current / 2, y),
+      targetHeight.current / zoom.current / 2
     )
 
-    // we don't multiply nX and nY by zoom.current here since image scale does that for us
-    cutImage.current.style.clipPath = makeSelector(imgW / 2 + nX, imgH / 2 + nY)
+    cutImage.current.style.clipPath = makeSelector(nX, nY)
 
-    // here we are in absolute coordinates, we have to multiply image offset by zoom
-    const imgX = (ctxW - imgW) / 2 - nX * zoom.current
-    const imgY = (ctxH - imgH) / 2 - nY * zoom.current
+    // we rotate and scale around cursor
+    const transformOriginValue = `${nX}px ${nY}px`
 
-    const transformValue = `translate(${imgX}px, ${imgY}px) scale(${zoom.current})`
+    // now we compensate for origin with -nX, -nY, rotate and scale
+    const imgX = container.current.clientWidth / 2 - nX
+    const imgY = container.current.clientHeight / 2 - nY
+    const transformValue = `translate(${imgX}px, ${imgY}px) rotate(${
+      rotation.current
+    }deg) scale(${zoom.current * flipDirX.current}, ${
+      zoom.current * flipDirY.current
+    })`
 
     cutImage.current.style.transform = transformValue
     fullImage.current.style.transform = transformValue
+    cutImage.current.style.transformOrigin = transformOriginValue
+    fullImage.current.style.transformOrigin = transformOriginValue
 
     return [nX, nY]
   }
 
+  // (x, y) is the point, (px, py) is the pivot
+  const rotate = (
+    x: number,
+    y: number,
+    angle: number,
+    px: number = 0,
+    py: number = 0
+  ) => {
+    const cos = Math.cos(angle)
+    const sin = Math.sin(angle)
+    x -= px
+    y -= py
+    return [cos * x - sin * y + px, sin * x + cos * y + py]
+  }
+
   useLayoutEffect(() => {
     const onMouseDown = (ev: MouseEvent) => {
       ev.preventDefault()
 
       dragging.current = true
 
-      startX.current = ev.clientX
-      startY.current = ev.clientY
+      // startX.current = ev.clientX
+      // startY.current = ev.clientY
+      ;[startX.current, startY.current] = rotate(
+        ev.clientX,
+        ev.clientY,
+        -(rotation.current * Math.PI) / 180,
+        window.innerWidth / 2,
+        window.innerHeight / 2
+      )
+    }
+
+    const handleMouseCoords = (ev: MouseEvent) => {
+      // const dx = ev.clientX, dy = ev.clientY
+      const [dx, dy] = rotate(
```

Comment:
These `dx` and `dy` look gorgeous xD

Reply:
> These `dx` and `dy` look gorgeous xD
> 
> (Edit: it was from mathematician point of view!)

Told you, I lost my mind debugging this :D Thanks!